pkg/xds/experimental/client/listener.go

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package client
+
+import (
+	"log/slog"
+
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// listener is a helper structure for watchers focused on handling a single, registered callback.
+type listener struct {
+	ID      uint64
+	typeUrl string
+	cb      WatcherCallback
+	trigger chan uint64
+
+	log       *slog.Logger
+	resources xds.ResourceSource
+
+	lastVersion uint64
+	mux         lock.Mutex
+	nextVersion uint64
+}
+
+var _ xds.ResourceVersionObserver = (*listener)(nil)
+
+// HandleNewResourceVersion implements xds.ResourceVersionObserver.
+// It triggers the callback process when a new version appears.
+func (l *listener) HandleNewResourceVersion(typeUrl string, version uint64) {
+	if typeUrl != l.typeUrl {
+		return
+	}
+	l.mux.Lock()
+	defer l.mux.Unlock()
+	if version <= l.lastVersion {
+		return
+	}
+	l.nextVersion = version
+	select {
+	case l.trigger <- version:
```

Comment:
Something about this feels racy to me. Does the receiver on l.trigger chan look at `version` that is sent?

Reply:
no

Reply:
Perhaps it would make sense to change the type of `l.trigger` from `chan unit64` to just `chan struct{}` like I suggested when reviewing this internally?

Reply:
Cool, thanks, I see it is applied now. Apparently, I was looking at the earlier version on which Bowei commented.